rust/datafusion/src/execution/context.rs

```diff
@@ -255,26 +253,17 @@ impl ExecutionContext {
 
     /// Register a table using a custom TableProvider so that it can be referenced from SQL
     /// statements executed against this context.
-    pub fn register_table(
-        &mut self,
-        name: &str,
-        provider: Box<dyn TableProvider + Send + Sync>,
-    ) {
-        let mut ctx_state = self.state.lock().expect("failed to lock mutex");
-        ctx_state.datasources.insert(name.to_string(), provider);
+    pub fn register_table(&mut self, name: &str, provider: Box<dyn TableProvider>) {
```

Comment:
The table provider is used during single-thread planning (not multi-threaded execution) and thus there is no reason for the table provider to be `Send` and `Sync` anymore

Reply:
I think that we aimed for `TableProvider` to be `Send + Sync` on ARROW-9425 , so that others could use `ExecutionContext` on multi-threaded environments such as Python.

See e.g. the usage [here](https://github.com/jorgecarleitao/datafusion-python/blob/master/src/lib.rs#L57). I think that that code only compiles if `ExecutionContext` is sharable between threads.

Reply:
I agree that `TableProvider` no longer needs `Send` and `Sync` now that it is just used during planning and no longer performs execution.

I don't think changing this will affect whether ExecutionContext can be sent between threads but I don't have time today to actually confirm that.

If we do want to be sure that `ExecutionContext` can be shared between threads, we should add a unit test for that.

Hopefully, I'll be able to find some time this evening to look at this some more.

Reply:
I was not able to find time to look at this. @jorgecarleitao Could you summarize the use case for sharing an ExecutionContext betweeen threads?

Reply:
TL;DR: let's ship this, as master is already not thread safe.

The two use-cases I though at the time:

1. when a project is planning a complex number of plans that depend on a common set of sources and UDFs, it would be nice to be able to multi-thread the planning. This is particularly important when planning requires reading remote metadata to formulate themselves (e.g. when the source is in `s3` with many partitions). Metadata reading is often slow and network bounded, which makes threads suitable for these workloads. If multi-threading is not possible, either each plan needs to read the metadata independently (one context per plan) or planning must be sequential (with lots of network waiting).

2. when creating bindings to programming languages that support multi-threading, it would be nice for the `ExecutionContext` to be thread safe, so that we can more easily integrate with those languages.

The reason for 2 is described in [`pyo3`'s documentation](https://pyo3.rs/master/class.html#defining-a-new-class):

> Because Python objects are freely shared between threads by the Python interpreter, all structs annotated with #[pyclass] must implement Send

Thread safety is not a strict requirement, though `pyo3` offers the option to `panic!` if the instance is used across threads. Not a great user experience, but definitely better than undefined behavior :)

The code below works in DataFusion 1.0.0, and no longer works in master:

```
#[pyclass]
struct ExecutionContext {
    ctx: datafusion::execution::context::ExecutionContext,
}
```

> `(dyn datafusion::execution::physical_plan::PhysicalPlanner + 'static)` cannot be shared between threads safely

Since this is already an issue on master, I think that we just ship this, that is a great simplification.

However, I think that we would enjoy some clarity of our stance wrt to thread safety of the `ExecutionContext`.

Reply:
I tried to capture the desire for ExecutionContext to be shared across threads in this ticket:  https://issues.apache.org/jira/browse/ARROW-9888 (and it has a test that we could add to ensure that execution context remains shareable across threads, if we want to go that route)

Reply:
I'll get a PR up to fix the thread safety issue shortly

Reply:
https://github.com/apache/arrow/pull/8082/